File: NET/API/Treehopper.Android/UsbConnection.cs
using System;
using System.ComponentModel;
using System.Threading.Tasks;
using Android.App;
using Android.Content;
using Android.Hardware.Usb;
using Java.Lang;

/// <summary>
/// Xamarin-based implementation for running Treehopper on Android
/// </summary>
namespace Treehopper.Android
{
    /// <summary>
    /// UsbConnection implementation for Xamarin.Android
    /// </summary>
    public class UsbConnection : IConnection
    {
        UsbDeviceConnection connection;
        readonly UsbManager usbManager;
        readonly UsbDevice usbDevice;

        UsbEndpoint pinConfigEndpoint;
        UsbEndpoint pinReportEndpoint;
        UsbEndpoint peripheralConfigEndpoint;
        UsbEndpoint peripheralResponseEndpoint;

        Thread pinListenerThread;
        bool pinListenerThreadRunning = false;

        public UsbConnection(UsbDevice device, Context appContext) : this(device, (UsbManager) appContext.GetSystemService("usb"))
        {

        }

        public UsbConnection(UsbDevice device, UsbManager manager)
        {
            usbDevice = device;
            usbManager = manager;
        }

        public string DevicePath => usbDevice.Handle.ToString();

        public string Name => usbDevice.ProductName;

        public string Serial => usbDevice.SerialNumber;

        public int UpdateRate { get; set; }

        string IConnection.DevicePath
        {
            get
            {
                throw new NotImplementedException();
            }
        }

        public short Version { get; private set; }

        public event PinEventData PinEventDataReceived;
        public event PropertyChangedEventHandler PropertyChanged;

        private bool connected = false;

        public void Close()
        {
            if (!connected)
                return;

            pinListenerThreadRunning = false;
            pinListenerThread.Join();

            connection.Close();
            connected = false;
        }

        public async Task<bool> OpenAsync()
        {
            if (connected)
                return false;

            //usbManager.RequestPermission(usbDevice, PendingIntent.GetBroadcast(ConnectionService.Instance.Context, 0, new Intent(ConnectionService.Instance.ActionUsbPermission), PendingIntentFlags.CancelCurrent));
            
            UsbInterface intf = usbDevice.GetInterface(0);
            pinReportEndpoint = intf.GetEndpoint(0);
            peripheralResponseEndpoint = intf.GetEndpoint(1);
            pinConfigEndpoint = intf.GetEndpoint(2);
            peripheralConfigEndpoint = intf.GetEndpoint(3);
            connection = usbManager.OpenDevice(usbDevice);
            if (connection != null)
            {
                bool intfClaimed = connection.ClaimInterface(intf, true);
                if (intfClaimed)
                {
                    connected = true;

                    pinListenerThread = new Thread(
                        () =>
                        {
                            byte[] data = new byte[64];
                            pinListenerThreadRunning = true;
                            while (pinListenerThreadRunning)
                            {
                                int res =
                                    connection.BulkTransfer(pinReportEndpoint, data, 41,
                                        100); // pin reports are 41 bytes long now
                                if (res > 0)
                                    this.PinEventDataReceived?.Invoke(data);
                            }
                        });

                    pinListenerThread.Start();
                    return true;

                }
            }
            return false;
        }

        public async Task<byte[]> ReadPeripheralResponsePacketAsync(uint bytesToRead)
        {
            byte[] data = new byte[bytesToRead];
            int res = connection.BulkTransfer(peripheralResponseEndpoint, data, (int)bytesToRead, 1000);
            return data;
        }

        public async Task SendDataPeripheralChannelAsync(byte[] data)
        {
            if (connection == null)
            {
                connected = false;
                return;
            }

            connection.BulkTransfer(peripheralConfigEndpoint, data, data.Length, 1000);
        }

        public async Task SendDataPinConfigChannelAsync(byte[] data)
        {
            if (connection == null)
            {
                connected = false;
                return;
            }

            connection.BulkTransfer(pinConfigEndpoint, data, data.Length, 1000);
        }

        public void Dispose()
        {
            
        }
    }
}

File: Java/api/treehopper.android/src/main/java/io/treehopper/UsbConnection.java
package io.treehopper;

import android.content.Context;
import android.hardware.usb.*;

import io.treehopper.*;
import io.treehopper.interfaces.Connection;

/**
 * Android Connection implementation
 */
public class UsbConnection implements Connection {
    private UsbDeviceConnection connection;
    private UsbManager usbManager;
    UsbDevice usbDevice;

    UsbEndpoint pinConfigEndpoint;
    UsbEndpoint pinReportEndpoint;
    UsbEndpoint peripheralConfigEndpoint;
    UsbEndpoint peripheralResponseEndpoint;

    Thread pinListenerThread;
    boolean pinListenerThreadRunning = false;
    private TreehopperUsb board;

    public UsbConnection(UsbDevice device, Context appContext) {
        this(device, (UsbManager) appContext.getSystemService(appContext.USB_SERVICE));
    }

    public UsbConnection(UsbDevice device, UsbManager manager) {
        usbDevice = device;
        serialNumber = device.getSerialNumber();
        name = device.getProductName();
        usbManager = manager;
    }

    public boolean open() {
        if (connected)
            return true;


        UsbInterface intf = usbDevice.getInterface(0);
        pinReportEndpoint = intf.getEndpoint(0);
        peripheralResponseEndpoint = intf.getEndpoint(1);
        pinConfigEndpoint = intf.getEndpoint(2);
        peripheralConfigEndpoint = intf.getEndpoint(3);
        connection = usbManager.openDevice(usbDevice);
        if (connection != null) {
            boolean intfClaimed = connection.claimInterface(intf, true);
            if (intfClaimed) {
                connected = true;

                pinListenerThread = new Thread() {
                    @Override
                    public void run() {
                        byte[] data = new byte[64];
                        pinListenerThreadRunning = true;
                        while (pinListenerThreadRunning) {
                            int res = connection.bulkTransfer(pinReportEndpoint, data, 41, 100); // pin reports are 41 bytes long now
                            if (board != null && res > 0)
                                board.onPinReportReceived(data);
                        }
                    }
                };

                pinListenerThread.start();
            }
        }

        return connected;
    }

    boolean connected;

    public boolean isConnected() {
        return connected;
    }

    public void close() {
        if (!connected)
            return;

        pinListenerThreadRunning = false;
        try {
            pinListenerThread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        connection.close();
        connected = false;
    }

    public String getSerialNumber() {
        return serialNumber;
    }

    String serialNumber;

    public String getName() {
        return name;
    }

    String name;

    public void sendDataPinConfigChannel(byte[] data) {
        if (connection == null) {
            connected = false;
            return;
        }

        int res = connection.bulkTransfer(pinConfigEndpoint, data, data.length, 1000);
    }

    public void sendDataPeripheralChannel(byte[] data) {
        if (connection == null) {
            connected = false;
            return;
        }

        int res = connection.bulkTransfer(peripheralConfigEndpoint, data, data.length, 1000);
    }

    @Override
    public byte[] readPeripheralResponsePacket(int numBytesToRead) {
        byte[] data = new byte[numBytesToRead];
        int res = connection.bulkTransfer(peripheralResponseEndpoint, data, numBytesToRead, 1000);
        return data;
    }

    @Override
    public void setPinReportListener(TreehopperUsb board) {
        this.board = board;
    }

}
